src/libexpr/primops/fetchClosure.cc

```diff
@@ -74,86 +190,87 @@ static void prim_fetchClosure(EvalState & state, const PosIdx pos, Value * * arg
+            fromStore = "https://cache.nixos.org";
+            fromPath = /nix/store/ldbhlwhh39wha58rm61bkiiwm6j7211j-git-2.33.1;
+        }
+        ```
+
+      - Fetch an [input-addressed store path](@docroot@/glossary.md#gloss-input-addressed-store-object) as is. This depends on user configuration, which is less preferable.
 
-      fetches `/nix/store/r2jd...` from the specified binary cache,
+        ```nix
+        builtins.fetchClosure {
+            fromStore = "https://cache.nixos.org";
+            fromPath = /nix/store/r2jd6ygnmirm2g803mksqqjm4y39yi6i-git-2.33.1;
+            inputAddressed = true;
+        }
+        ```
+
+      **Rewriting a store path**
+
+      This first example fetches `/nix/store/r2jd...` from the specified binary cache,
       and rewrites it into the content-addressed store path
       `/nix/store/ldbh...`.
 
-      If `fromPath` is already content-addressed, or if you are
-      allowing impure evaluation (`--impure`), then `toPath` may be
-      omitted.
+      By rewriting the store paths, you can add the contents to any store without requiring that you or perhaps your users configure any extra trust.
 
       To find out the correct value for `toPath` given a `fromPath`,
-      you can use `nix store make-content-addressed`:
+      you can use [`nix store make-content-addressed`](@docroot@/command-ref/new-cli/nix3-store-make-content-addressed.md):
 
       ```console
       # nix store make-content-addressed --from https://cache.nixos.org /nix/store/r2jd6ygnmirm2g803mksqqjm4y39yi6i-git-2.33.1
       rewrote '/nix/store/r2jd6ygnmirm2g803mksqqjm4y39yi6i-git-2.33.1' to '/nix/store/ldbhlwhh39wha58rm61bkiiwm6j7211j-git-2.33.1'
       ```
 
-      This function is similar to `builtins.storePath` in that it
+      Alternatively, you may set `toPath = ""` and find the correct `toPath` in the error message.
+
+      **Not rewriting**
+
+      `toPath` may be omitted when either
+       - `fromPath` is already content-addressed, or
+       - `inputAddressed = true;` is set.
+
+      Fetching an [input addressed path](@docroot@/glossary.md#gloss-input-addressed-store-object) is not recommended because it requires that the source be trusted by the Nix configuration.
+
+      **`builtins.storePath`**
+
+      This function is similar to [`builtins.storePath`](#builtins-storePath) in that it
       allows you to use a previously built store path in a Nix
       expression. However, it is more reproducible because it requires
       specifying a binary cache from which the path can be fetched.
-      Also, requiring a content-addressed final store path avoids the
-      need for users to configure binary cache public keys.
+      Also, the default requirement of a content-addressed final store path
+      avoids the need for users to configure binary cache public keys.
+
+      **Experimental feature**
 
       This function is only available if you enable the experimental
       feature `fetch-closure`.
```

Comment:
```suggestion
      This function is only available if you enable the experimental
      feature `fetch-closure`.
```
I think we shouldn't have pseudo-section-headings in those listings. They should not be all that long to begin with, and in this case you get enough structure from lists and paragraphs. For this part specifically, maybe we should first merge https://github.com/NixOS/nix/pull/8371 and implement https://github.com/NixOS/nix/issues/8466?

Reply:
We should have headers so that readers can skip to the information they care about. I don't care that we failed to put primops on their own pages.

Reply:
By removing this header, the paragraph about the experimental feature would become part of the `builtins.storePath` explanation.

Reply:
Eh, yeah. But then again we can also put the callout first. With the automatic callout PR we can restructure it as we like. And then even putting primops on separate pages is not a big deal, although I do think there isn't all that much to say about each of them. In any case it's an entirely independent issue.